### PR TITLE
fix: power Sound & Light before child controls

### PR DESCRIPTION
**Pull request by gatesry**
## Summary

Fixes Sound & Light child controls when master power is off. Home Assistant now powers the Sound & Light Machine on before sending light or sound commands.

The light and sound entities show as off while master power is off, matching what is actually visible or audible.

## Testing

Tested on my real Home Assistant setup with a Nanit Sound & Light Machine. I installed the updated HACS custom component files, restarted HA Core, and verified Nanit loaded.

On the live device, enabling light or sound powered the machine on first, then enabled the requested feature. Unit coverage was added.

### CODE REVIEW

**Comment by wealthystudent** [comment]
thanks for submitting!

i will review this in the comming days

File: custom_components/nanit/entity.py [hunk 1]
```diff
@@ -85,6 +85,14 @@ def available(self) -> bool:
...
         return self.coordinator.last_update_success and self.coordinator.data is not None
 
+    async def _async_power_on_sound_light(self) -> None:
+        """Power on the Sound & Light Machine before enabling child features.
+
+        Always send the command before child controls because coordinator
+        state can be stale after the device is toggled outside Home Assistant.
+        """
+        await self.coordinator.sound_light.async_set_power(True)
+
```
Comment (wealthystudent) [inline custom_components/nanit/entity.py:95]: **[Nit]** This method is inherited by `NanitSLPowerSwitch` which doesn't use it. Not a real problem (the alternative is duplicating the method in both `NanitSoundLightLight` and `NanitSLSoundSwitch`), but worth being aware of.

Also, the "always send" approach means every light/sound turn-on sends **two WebSocket commands** (power + child). This is the right reliability tradeoff, but worth confirming the S&L device handles rapid sequential commands well — e.g., does it ACK the power-on before the child command arrives, or could the child command be silently dropped if it arrives before power-on completes on the firmware side?

File: custom_components/nanit/light.py [hunk 2]
```diff
@@ -224,10 +224,12 @@ def __init__(
 
     @property
     def is_on(self) -> bool | None:
-        """Return True if the light is on."""
+        """Return True if the light is effectively on."""
         if self.coordinator.data is None:
             return None
         state = self.coordinator.data
+        if state.power_on is False:
```
Comment (wealthystudent) [inline custom_components/nanit/light.py:231]: **[Good]** `is False` identity check is correct here — when `power_on is None` (unknown/never-received), the entity falls through to the existing `light_enabled` / `brightness` logic. This preserves backward compatibility and avoids showing the light as "off" before the first power state is received.

**[Question]** When `power_on` transitions from `True` to `False` (e.g., user toggles the S&L power switch), this entity will transition from `on` → `off` even if `light_enabled` is still `True` in state. This is correct UX (the light is physically not visible), but could trigger automations listening for `light.turn_off` events. Worth a note in the PR description for future reference.

File: custom_components/nanit/light.py [hunk 3]
```diff
@@ -265,6 +267,7 @@ def hs_color(self) -> tuple[float, float] | None:
     async def async_turn_on(self, **kwargs: Any) -> None:
         """Turn on the light, optionally setting brightness and/or color."""
         try:
+            await self._async_power_on_sound_light()
```
Comment (wealthystudent) [inline custom_components/nanit/light.py:270]: **[Medium]** If `_async_power_on_sound_light()` throws `NanitTransportError`, the `except` block below catches it and logs `"Failed to control Sound & Light light"` — which is accurate but ambiguous. From the log alone, you can't tell whether the **power-on step** or the **light-enable step** failed.

For debugging field issues, consider either:
- Wrapping the power-on call in its own try/except with a specific log message, or
- Enriching the error message to include which sub-step failed

This also applies to the sound switch in `switch.py`.

```python
# Example approach (optional):
try:
    await self._async_power_on_sound_light()
except NanitTransportError as err:
    _LOGGER.error("Failed to power on S&L before enabling light: %s", err)
    return
try:
    await self.coordinator.sound_light.async_set_light_enabled(True)
    # ... color, brightness ...
except NanitTransportError as err:
    _LOGGER.error("Failed to control Sound & Light light: %s", err)
```

File: custom_components/nanit/switch.py [hunk 4]
```diff
@@ -248,15 +248,18 @@ def __init__(
 
     @property
     def is_on(self) -> bool | None:
-        """Return True if sound is on."""
+        """Return True if sound is effectively on."""
         if self.coordinator.data is None:
             return None
```
Comment (wealthystudent) [inline custom_components/nanit/switch.py:253]: **[Good]** Consistent with the light entity pattern — same `is False` identity check, same fallback semantics.

File: custom_components/nanit/switch.py [hunk 5]
```diff
@@ -248,15 +248,18 @@ def __init__(
...
+        if self.coordinator.data.power_on is False:
+            return False
         result: bool | None = self.coordinator.data.sound_on
         return result
 
     async def async_turn_on(self, **kwargs: Any) -> None:
         """Turn sound on."""
         try:
+            await self._async_power_on_sound_light()
```
Comment (wealthystudent) [inline custom_components/nanit/switch.py:262]: **[Low]** Same error ambiguity concern as `light.py` — if `_async_power_on_sound_light()` fails, the log says `"Failed to turn on S&L sound"` when it was actually the power-on that failed.

Also a subtle state consideration: if the power command succeeds but `async_set_sound_on` fails (e.g., transient network issue between the two commands), the device ends up powered on but with no sound playing. The S&L power switch entity will correctly show "on" (since `async_set_power` updates state optimistically), but the sound entity won't call `async_write_ha_state()`. The coordinator's `STATE_UPDATE` event from the power command should trigger a re-render, so this is probably fine — just documenting the thought.

File: tests/unit/test_sl_entities.py [hunk 6]
```diff
@@ -169,6 +183,44 @@ async def test_light_turn_on_calls_set_light_enabled_true() -> None:
...
+    calls: list[str] = []
+    coordinator.sound_light.async_set_power.side_effect = lambda _on: calls.append("power")
+    coordinator.sound_light.async_set_light_enabled.side_effect = lambda _on: calls.append("light")
+    entity = NanitSoundLightLight(coordinator)
+    _disable_state_writes(entity)
+
+    await entity.async_turn_on()
+
+    assert calls[:2] == ["power", "light"]
+    coordinator.sound_light.async_set_power.assert_awaited_once_with(True)
```
Comment (wealthystudent) [inline tests/unit/test_sl_entities.py:201]: **[Good]** Solid test suite addition. Particularly like:
- `test_light_turn_on_applies_power_enable_color_brightness_in_order` — verifies the exact command sequence
- `test_light_turn_on_does_not_enable_light_when_power_on_fails` — confirms failure propagation stops child commands
- `test_light_turn_off_does_not_turn_off_device_power` — explicitly documents the intentional asymmetry
- `test_light_turn_on_sends_power_command_even_when_state_already_powered` — validates the "always send" design choice

**Review by wealthystudent — COMMENTED** [review]
## PR Review: fix: power Sound & Light before child controls

### Summary

Clean, well-scoped PR that solves a real usability problem. The approach of always sending the power-on command before child controls is pragmatic for IoT reliability where coordinator state can be stale. Test coverage is thorough — ordering, failure propagation, power-already-on, unknown-power, and turn-off-doesn't-touch-power are all covered.

I ran `just check` and `just test` locally on the PR branch. All lint/format checks pass, 344 tests pass (the 4 errors are pre-existing on `main` — lingering timer/task issues in `test_sl_coordinator.py` and `test_sound_light.py`).

### What's good

- **`is False` identity check** — correctly preserves `None` fallback behavior so entities remain backward-compatible when `power_on` is unknown
- **Base class method placement** — `_async_power_on_sound_light()` on `NanitSoundLightEntity` avoids duplication between light and sound entities
- **"Always send" power approach** — the right tradeoff for an IoT device where stale state is common. The docstring explains the rationale well
- **Turn-off doesn't touch device power** — correct, since the other feature (sound/light) may still be active
- **16 new passing tests** with good edge case coverage

### Observations (see inline comments)

1. **[Medium]** Error log ambiguity when power-on step fails inside `async_turn_on`

Overall: good to merge with minor suggestions. None of the observations are blockers.

**Comment by wealthystudent** [comment]
just look at the` _async_power_on_sound_light()` debug improvement i listed above, then it is ready

**Comment by gatesry** [comment]
> just look at the` _async_power_on_sound_light()` debug improvement i listed above, then it is ready

Sounds good! Thanks for the quick review. Been running the changes a few days on the S&L device. Will check this out and submit.

**Comment by wealthystudent** [comment]
Hey @gatesry, really sorry about closing this. Your fix was solid and the review feedback was already addressed, so this is not about the quality of your work at all. I'm doing a larger S&L refactoring (#102) that will rewrite how S&L controls work. We're merging in a dedicated S&L integration ([nanit-sound-light](https://github.com/com6056/nanit-sound-light)) that has a more complete transport model, so merging this now would just create conflicts down the line.

The power-on-before-child-controls behavior will absolutely be included in the new architecture. I really appreciate you taking the time to contribute and test on real hardware, and I hope you'll keep contributing when the new S&L code lands. Thanks again 🙂

**Comment by gatesry** [comment]
Thanks for letting me new about the new S&L integration!
